nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -645,8 +645,99 @@ impl StringPrototype {
         ))
     }
 
-    fn replace(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
+    /// ### [22.1.3.19 String.prototype.replace ( searchValue, replaceValue )](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replace)
+    fn replace(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+
+        let search_value = args.get(0);
+        let replace_value = args.get(1);
+
+        // Early path: use rust's replace if everything are strings: `"foo".replace("o", "a")`
+        if let (Ok(s), Ok(search_string), Ok(replace_string)) = (
+            String::try_from(o),
+            String::try_from(search_value),
+            String::try_from(replace_value),
+        ) {
+            let result = s
+                .as_str(agent)
+                .replace(search_string.as_str(agent), replace_string.as_str(agent));
+            return Ok(String::from_string(agent, result).into_value());
+        }
+
+        // 2. If searchValue is neither undefined nor null, then
+        if !search_value.is_null() && !search_value.is_undefined() {
+            // a. Let replacer be ? GetMethod(searchValue, %Symbol.replace%).
+            let symbol = WellKnownSymbolIndexes::Replace.into();
+            let replacer = get_method(agent, search_value, symbol);
+
+            // b. If replacer is not undefined, Return ? Call(replacer, searchValue, « O, replaceValue »).
+            if let Ok(Some(replacer)) = replacer {
+                return call_function(
+                    agent,
+                    replacer,
+                    search_value,
+                    Some(ArgumentsList(&[o, replace_value])),
+                );
+            }
+        }
+
+        // 3. Let s be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 4. Let searchString be ? ToString(searchValue).
+        let search_string = to_string(agent, search_value)?;
+
+        // 5. Let functionalReplace be IsCallable(replaceValue).
+        match is_callable(replace_value) {
```

Comment:
**nitpick:** I'd do this as
```rs
if let Some(functional_replace) = is_callable(replace_value) {
 // Some branch here, step 7 etc. Ends with a return.
}
// None branch but now with less None matching visible
let replace_string = ...;
```
But this way is of course _absolutely_ fine. No need to change, this is just me nitpicking for the sake of it I guess :D